### ISSUE TITLE
ndt7: experiment with larger messages

### ISSUE THREAD

**Issue by bassosimone** [m-lab/ndt-server#130]
During a chat with Domotz it emerged that their ndt7 clients have issues processing the download messages because they receive too many messages. Hence, it would be interesting to see whether using larger messages solve their use case.

**Comment by bassosimone** [comment]
I've run an experiment with a server with `1<<20` messages on a Docker container and a headless Chrome using https://github.com/bassosimone/ndt7-puppeteer on another Docker.

This is an excerpt from the results:

```json
{  
   "origin":"server",
   "measurement":{  
      "bbr_info":{  
         "max_bandwidth":1943115352,
         "min_rtt":0.129
      },
      "elapsed":6.500166806,
      "tcp_info":{  
         "smoothed_rtt":0.617,
         "rtt_var":0.288
      }
   }
}
```

This is the same experiment using `1<<13` (the value we currently use):

```json
{  
   "origin":"server",
   "measurement":{  
      "bbr_info":{  
         "max_bandwidth":984015256,
         "min_rtt":0.147
      },
      "elapsed":2.25017943,
      "tcp_info":{  
         "smoothed_rtt":2.016,
         "rtt_var":0.61
      }
   }
}
```

The interesting bit is that with larger messages we see 1.9 Gbit/s of max bandwidth and with the standard messages we see 0.9 Gbit/s of bandwidth.

**Comment by bassosimone** [comment]
@Giannandrea has run some tests towards the same server using a constrained client for which the CPU was quite likely an issue and confirmed that larger messages helped his use case.

**Comment by pboothe** [comment]
Please try with `1<<24` message sizes as well.

Every received message causes the JS event loop to fire. The JS event loop can only fire so fast, especially inside a browser.

**Comment by bassosimone** [comment]
Here's another experiment. With `1<<13` between CGE fra and GCE ams:

```json
{  
   "origin":"server",
   "measurement":{  
      "bbr_info":{  
         "max_bandwidth":142447448,
         "min_rtt":7.345
      },
      "elapsed":5.250210045,
      "tcp_info":{  
         "smoothed_rtt":8.078,
         "rtt_var":0.398
      }
   }
}
```

That's 142 Mbit/s.

With `1<<20`:

```json
{  
   "origin":"server",
   "measurement":{  
      "bbr_info":{  
         "max_bandwidth":831846128,
         "min_rtt":7.388
      },
      "elapsed":4.750226899,
      "tcp_info":{  
         "smoothed_rtt":7.854,
         "rtt_var":0.228
      }
   }
}
```

That's 831 Mbit/s.

With `1<<24`:

```json
{  
   "origin":"server",
   "measurement":{  
      "bbr_info":{  
         "max_bandwidth":1178109800,
         "min_rtt":7.405
      },
      "elapsed":6.000238165,
      "tcp_info":{  
         "smoothed_rtt":7.601,
         "rtt_var":0.054
      }
   }
}
```

That's 1.17 Gbit/s.

**Comment by pboothe** [comment]
I think we need to leave the door open for very large messages in case we encounter clients with bad CPUs and good network connections.

**Comment by bassosimone** [comment]
I think I need to think at how to select large messages. It's unclear to me whether a client could flag itself has being in this particular condition or the server should guess.

**Comment by mattmathis** [comment]
We really need some .pcaps.  Although larger messages get higher average performance, I be the receiver completely stallsTCP, because it is CPU bound.   We may need to revisit client certifications too.

I also want to confirm that it is properly accounted in walltime minus tcp-info busytime.

**Comment by pboothe** [comment]
The client should not flag itself. Instead, we should figure out how to gradually ramp up the message size when we get evidence that the number of messages per second is above some threshold value.

**Comment by bassosimone** [comment]
Yesterday we collected another data point confirming that the browser client implementation is slower than the Go client implementation with 1 Gbit/s networks. It should be interesting to see whether setting larger messages is helpful in that use case as well (I expect this to happen).

**Comment by bassosimone** [comment]
At the same time, it seems that the message size increase does not solve all Domotz issues. I am going to talk more to @Giannandrea and see if we can understand the root cause.

**Comment by bassosimone** [comment]
The following is a plot of BBR's Max Bandwidth when changing the WebSocket message sizes in tests run between GCE's Frankfurt and GCE's Amsterdam zones, using v0.9.0 of the server and using the JavaScript client it embeds running on Google Chrome instrumented using [puppeteer](https://github.com/GoogleChrome/puppeteer). Both the client and the server were inside Docker.

The following is the same experiment where instead we're using https://github.com/m-lab/ndt7-client-go/commit/f900c9f862c2e97ee53372f4311a3ea35fb3cb3c:

![plot](https://user-images.githubusercontent.com/337298/59145795-feb1e480-89e8-11e9-9009-3cde1321d6dd.jpg)

**Comment by bassosimone** [comment]
The following is the same experiment where instead we're using https://github.com/m-lab/ndt7-client-go/commit/f900c9f862c2e97ee53372f4311a3ea35fb3cb3c:

![Go client](https://user-images.githubusercontent.com/337298/59145795-feb1e480-89e8-11e9-9009-3cde1321d6dd.jpg)

**Comment by bassosimone** [comment]
With @mattmathis we concluded that the clients were using 100% of CPU in the Domotz case and we have yet to take another look at what limits performance in cloud tests at around 3 Gbit/s.

Meanwhile, with @Giannandrea, we understood that `armv7` does not implement specific instructions to make encryption faster and we have seen (see https://github.com/golang/go/issues/18498) that Go performance dramatically improved when asm code was added for `armv8`. Also, apparently, the `armv7` architecture does not support such instructions.

I believe the issue discussed with @Giannandrea is more related to this later factor than it is related with larger/smaller messages; and, in fact, the message size played a dubious role in our tests. So, this is most likely a separate issue from the one discussed in here (even though at the beginning I was quite sure the two issues, cloud and armv7 devices, had the same root cause.)

**Comment by bassosimone** [comment]
I think this has lower priority than `P0`. I am tentatively adding `P2`. The main problem to be solved here is how to change the message size such that fast browser clients work well.

**Comment by bassosimone** [comment]
With @pboothe we've determined that a good solution to this problem that does not involve complicating the protocol is to gradually increase the binary message size as the download subtest is making progress. Limited testing showed that this solution is reasonable for both the Go and the JavaScript clients. What's more, it seems using `1<<20` as opposed to `1<<24` for the message is a good compromise for both clients, whereas `1<<24` optimises the performance of the JavaScript client by a small percentage but causes a ~30% performance drop for the Go client.

**Comment by bassosimone** [comment]
I believe https://github.com/m-lab/ndt-server/issues/130#issuecomment-526155692 is enough to call this research effort over. There should be a follow-up hands on issue pertaining how to properly  implement this functionality in this codebase.

**Comment by bassosimone** [comment]
The above solution has been documented in #180 and, furthermore, has been implemented inside of two research implementations (also mentioned in the spec):

- https://github.com/bassosimone/ndt7-server-go-minimal

- https://github.com/bassosimone/ndt7-client-go-minimal

Also, we have confirmed that `1<<20` seems to be a good compromise between the needs to both Go and JavaScript. It seems there is no gain in setting gorilla/websocket buffers to `1<<24`in terms of performance, even though that avenue look promising. Since these tests are in 1+ Gbit/s scenarios and there are many slower connections around, it makes sense to eventually look again into this topic later. Now it is probably more sensible to focus on making the ndt7 beta available.
